### PR TITLE
0.9.4

### PR DESCRIPTION
**Pull request by tbchrist**
* Upgrade helper improvements
 * Generalized lockless master promotion
 * Updated simple tracker
 * Utility methods to track query counters
 * Updates to physical locality concepts
 * Other miscellaneous improvements

### CODE REVIEW

**Comment by KirilAngov** [comment]
:bullettrain_front: LGTM

**Comment by dallasmarlow** [comment]
@bobpattersonjr seems like the `percona_schema` database references could be a config value, also hi! :ramen:

File: plugins/jetpants_collins/topology.rb [hunk 1]
```diff
@@ -88,8 +108,13 @@ def claim_spares(count, options={})
     def count_spares(options={})
       query_spare_assets(100, options).count
     end
-    
-    
+
+    # This method won't ever return more than than 100 nodes, but that's
+    # not a problem, since no single operation requires that many spares
+    def spares(options={})
+      query_spare_assets(100, options).map(&:to_db)
```
Comment (dallasmarlow) [inline plugins/jetpants_collins/topology.rb]: @KirilAngov `concurrent_map` instead?

File: plugins/jetpants_collins/topology.rb [hunk 2]
```diff
@@ -274,12 +313,43 @@ def query_spare_assets(count, options={})
...
-        if(db.usable_spare? && (!source || db.usable_with?(source)))
-          keep_nodes << node
-          break if keep_nodes.size >= count
+        if(db.usable_spare? &&
+          (
+            !source ||
+            (!source.pool && db.usable_with?(source)) ||
+            (
+              (!options[:for_pool] && source.pool && db.usable_in?(source.pool)) ||
+              (options[:for_pool] && db.usable_in?(options[:for_pool]))
```
Comment (dallasmarlow) [inline plugins/jetpants_collins/topology.rb]: shouldn't the options expressions be first?

Comment (KirilAngov) [inline plugins/jetpants_collins/topology.rb]: Would it make a difference in this case?

File: plugins/jetpants_collins/topology.rb (hunk 1 above)
Comment (KirilAngov) [inline plugins/jetpants_collins/topology.rb]: If `to_db` is not that slow, would it make a difference? I am really asking :)

File: plugins/jetpants_collins/topology.rb (hunk 2 above)
Comment (dallasmarlow) [inline plugins/jetpants_collins/topology.rb]: @KirilAngov it appears so, but in truth this conditional is a bit messy and could be broken up for clarity

Comment (tbchrist) [inline plugins/jetpants_collins/topology.rb]: Yup, this method in general is a little heavy right now and is a target to be cleaned up a bit in future iterations.

Specifically if we're looking at ``source`` then it's basically an alias of ``options[:like]`` (which is a bit confusing stylistically) and is also optional.  The stuff around ``options[:for_pool]`` is more or less longhand for the ``compare_pool logic`` elsewhere in the code which was written later, but will eventually be unified to be consistent.

**Pull request merged**